### PR TITLE
do not check regex optionally (closes #2074)

### LINKED ISSUE #2074: Remove regex check for test files
### Are you requesting a feature or reporting a bug?
request a feature

### What is the current behavior?
There are [regexes](https://github.com/DevExpress/testcafe/blob/master/src/compiler/test-file/api-based.js#L15) that each test file is run against, and if they do not match, you get an error

> No tests to run. Either the test files contain no tests or the filter function is too restrictive.

### What is the desired behavior?
Can this regex test be disabled and instead look for tests when actually running/parsing the code? I've found valid use cases where those regexes might not match.

In my project, I've broken up my app into several "subapp" libraries, and each subapp exports a suite of tests, which are run in the main repo's testcafe suite.

for example:

library test file
```js
export default function libraryTests(){
  fixture('Library tests').page('localhost:1234');

  test('Does some stuff', async t => {
    //await t.expect(....); //expect something
  });
}
```
main test file
```js
import {libraryTest} from 'library';

libraryTest();
```

currently, this fails when i try to run the tests pointing at the main test file. However, it works if i add a commented out fixture and test. Ideally, I would not need to add a commented out fixture and test just to satisfy the regex.

### PR DESCRIPTION
**Pull request by AlexKamaev**

### CODE REVIEW

File: test/server/compiler-test.js [hunk 1]
```diff
@@ -140,6 +173,39 @@ describe('Compiler', function () {
 
 
     describe('TypeScript', function () {
+        it('Should compile test defined in separate module if option is enabled', function () {
```
Comment (miherlosev) [inline test/server/compiler-test.js]: Why does we need addtional tests written with TypeScript and CoffeeScript?

File: test/server/compiler-test.js [hunk 2]
```diff
@@ -47,6 +47,39 @@ describe('Compiler', function () {
     });
 
     describe('ES-next', function () {
+        it('Should compile test defined in separate module if option is enabled', function () {
+            const sources = [
```
Comment (miherlosev) [inline test/server/compiler-test.js]: Also add tests for negative cases: 
If there are no `test` and/or `fixture` definitions into imported files, then TestCafe should report clear error messages.

Comment (AlexKamaev) [reply test/server/compiler-test.js]: ok

File: test/server/compiler-test.js (hunk 1 above)
Comment (AlexKamaev) [reply test/server/compiler-test.js]: We have several compilers and every on them has it's own set of similar tests. It's true that I modified base behavior which affects all of the compilers at the same time, but I think it'd better to have full set of tests

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

File: src/runner/index.js [hunk 3]
```diff
@@ -266,7 +266,7 @@ export default class Runner extends EventEmitter {
             .then(() => {
                 this._validateRunOptions();
 
-                return this.bootstrapper.createRunnableConfiguration();
+                return this.bootstrapper.createRunnableConfiguration({ disableTestSyntaxValidation });
```
Comment (miherlosev) [inline src/runner/index.js]: I propose to pass `disableTestSyntaxValidation` option via `this.bootstrapper.disableTestSyntaxValidation` as other bootstapper-related options such as `browsers`, `reporters` and etc.

File: src/compiler/test-file/base.js [hunk 4]
```diff
@@ -19,8 +19,8 @@ export default class TestFileCompilerBase {
         throw new Error('Not implemented');
     }
 
-    canCompile (code, filename) {
-        return this.supportedExtensionRe.test(filename) && this._hasTests(code);
+    canCompile (code, filename, disableTestSyntaxValidation) {
+        return this.supportedExtensionRe.test(filename) && (disableTestSyntaxValidation || this._hasTests(code));
```
Comment (AndreyBelym) [inline src/compiler/test-file/base.js]: You can just use `this.disableTestSyntaxValidation` here.

File: src/cli/argument-parser.js [hunk 5]
```diff
@@ -104,6 +104,8 @@ export default class CLIArgumentParser {
             .option('--dev', 'enables mechanisms to log and diagnose errors')
             .option('--qr-code', 'outputs QR-code that repeats URLs used to connect the remote browsers')
             .option('--sf, --stop-on-first-fail', 'stop an entire test run if any test fails')
+            .option('--disable-test-syntax-validation', 'disable test syntax validation')
```
Comment (AndreyBelym) [inline src/cli/argument-parser.js]: \cc @VasilyStrelyaev for approval

Comment (VasilyStrelyaev) [reply src/cli/argument-parser.js]: Well, we don't describe much here...
What if we say, for instance,
```js
            .option('--disable-test-syntax-validation', 'allows you to provide .js files that do not contain tests, while tests are imported or generated dynamically.')
```
(not checked with Dirk, just a suggestion)

File: src/compiler/test-file/base.js (hunk 4 above)
Comment (AlexKamaev) [reply src/compiler/test-file/base.js]: `this.disableTestSyntaxValidation` is defined in `Compiler` which is not based class for `EsNextTestFileCompiler`, `TypeScriptTestFileCompiler` etc.

File: src/cli/argument-parser.js (hunk 5 above)
Comment (miherlosev) [reply src/cli/argument-parser.js]: I agree with @VasilyStrelyaev. We can provide more detailed option description.
```js
disable `test` and `fixture` directives checking/validation (it allows to run dynamically loaded tests)
```

Comment (AlexKamaev) [reply src/cli/argument-parser.js]: I've commited with Vasily case for now, in any case we are waiting for Dirk

Comment (miherlosev) [reply src/cli/argument-parser.js]: Ok. Need to discuss with @VasilyStrelyaev

Comment (VasilyStrelyaev) [reply src/cli/argument-parser.js]: Dirk's version:

`disables checks for 'test' and 'fixture' directives to run dynamically loaded tests`

**Comment by VasilyStrelyaev** [comment]
https://github.com/DevExpress/testcafe/pull/2980#discussion_r227760489

File: src/cli/cli.js [hunk 6]
```diff
@@ -87,6 +87,9 @@ async function runTests (argParser) {
 
     reporters.forEach(r => runner.reporter(r.name, r.outStream));
 
+    if (opts.disableTestSyntaxValidation)
```
Comment (miherlosev) [inline src/cli/cli.js]: I propose to pass `disableTestSyntaxValidation` as a run option.

**Comment by AlexKamaev** [comment]
ping
